### PR TITLE
chore: Disallow cross-project reference lookups

### PR DESCRIPTION
**Pull request by tssweeney**
This PR disallows cross-project reference lookups. We implicitly enabled this feature here: https://github.com/wandb/weave/pull/1967, but for a number of reasons, we want to disable it.

Please see https://github.com/wandb/core/pull/23266 for additional commentary

### CODE REVIEW

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 1]
```diff
@@ -702,60 +712,77 @@ def refs_read_batch(self, req: tsi.RefsReadBatchReq) -> tsi.RefsReadBatchRes:
...
-                if len(conds) > 0:
-                    conditions = [combine_conditions(conds, "OR")]
-                    objs = self._select_objs_query(
-                        project_id, conditions=conditions, parameters=parameters
-                    )
-                    for obj in objs:
-                        root_val_cache[make_obj_cache_key(obj)] = json.loads(
-                            obj.val_dump
-                        )
+            for ref_index, ref in enumerate(refs):
```
Comment (tssweeney) [inline weave/trace_server/clickhouse_trace_server_batched.py:753]: Mostly an indentation change - hide whitespace for easier review

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 2]
```diff
@@ -897,7 +953,7 @@ def resolve_extra(extra: list[str], val: typing.Any) -> typing.Any:
                         extra_result.remaining_extra, extra_result.val
                     )
 
-        return tsi.RefsReadBatchRes(vals=[r.val for r in extra_results])
+        return [r.val for r in extra_results]
```
Comment (gtarpenning) [inline weave/trace_server/clickhouse_trace_server_batched.py:956]: why can we return just the list here?

Comment (gtarpenning) [reply weave/trace_server/clickhouse_trace_server_batched.py:956]: ah i see on line 736

**Review by gtarpenning — APPROVED** [review]
this change somehow makes this behemoth function (now two) easier to reason about. nice

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 3]
```diff
@@ -702,16 +704,43 @@ def refs_read_batch(self, req: tsi.RefsReadBatchReq) -> tsi.RefsReadBatchRes:
...
+            )
+            for ref, result in zip(project_refs, project_results):
+                final_result_cache[make_ref_cache_key(ref)] = result
+
+        # Return the final data payload
+        vals = [final_result_cache[make_ref_cache_key(ref)] for ref in parsed_refs]
+        return tsi.RefsReadBatchRes(vals=vals)
+
+    def _refs_read_batch_within_project(
+        self, project_id_scope: str, parsed_refs: ObjRefListType
```
Comment (gtarpenning) [inline weave/trace_server/clickhouse_trace_server_batched.py:739]: not 100% sure what the value is in changing `project_id` to `project_id_scope`

Comment (tssweeney) [reply weave/trace_server/clickhouse_trace_server_batched.py:739]: mainly for code readability - basically `project_id` is part of a `ref`, so it was a little confusing. `project_id_scope` just is a more clear noun

File: weave/trace_server/trace_server_converter.py [hunk 4]
```diff
@@ -45,8 +47,14 @@ def replace_ref(ref_str: str) -> str:
...
-        if isinstance(obj, str) and obj.startswith(weave_prefix):
-            return typing.cast(B, replace_ref(obj))
+        if isinstance(obj, str):
+            if obj.startswith(weave_prefix):
+                return typing.cast(B, replace_ref(obj))
+            elif obj.startswith(weave_internal_prefix):
+                # It is important to raise here as this would be the result of
+                # an external client attempting to write internal refs directly.
+                # We want to maintain full control over the internal refs.
+                raise ValueError("Encountered internal ref in external data.")
```
Comment (gtarpenning) [inline weave/trace_server/trace_server_converter.py:57]: Is there a reason we might want to obfuscate the reason to raise here? I can't actually think of a good reason, other than, if this only happens when someone is doing something they shouldn't, returning a handy error explaining why doesn't feel right.

Comment (tssweeney) [reply weave/trace_server/trace_server_converter.py:57]: Hmm - interesting. These will get transformed into 500s i believe, but really should be a 400.

Comment (tssweeney) [reply weave/trace_server/trace_server_converter.py]: ok, modified the tests to assert on status code and removed these messages.

**Pull request merged**